src/hotspot/share/opto/vectorization.cpp

```diff
@@ -1743,3 +1745,526 @@ void AlignmentSolver::trace_constrained_solution(const int C_const,
+          // and discover that use is also pre_visited but not post_visited. Thus, use
+          // lies on that path from "root" to vtn, and the edge (vtn, use) closes a
+          // circle.
+          NOT_PRODUCT(if (_is_trace_rejections) { trace_schedule_cycle(stack, pre_visited, post_visited); } )
+          return false;
+        }
+        stack.push(use);
+        all_uses_already_visited = false;
+      }
+
+      if (all_uses_already_visited) {
+        stack.pop();
+        post_visited.set(vtn->_idx);           // post-visit
+        _schedule.at_put_grow(rpo_idx--, vtn); // assign rpo_idx
+      }
+
+    } else {
+      stack.pop(); // Already post-visited. Ignore secondary edge.
+    }
+  }
+
+#ifndef PRODUCT
+  if (_is_trace_verbose) {
+    print_schedule();
+  }
+#endif
+
+  assert(rpo_idx == -1, "used up all rpo_idx, rpo_idx=%d", rpo_idx);
+  return true;
+}
+
+// Push all "root" nodes, i.e. those that have no inputs (req or dependency):
+void VTransformGraph::schedule_collect_nodes_without_req_or_dep(GrowableArray<VTransformNode*>& stack) const {
+  for (int i = 0; i < _vtnodes.length(); i++) {
+    VTransformNode* vtn = _vtnodes.at(i);
+    if (!vtn->has_req_or_dep()) {
+      stack.push(vtn);
+    }
+  }
+}
+
+#ifndef PRODUCT
+void VTransformGraph::trace_schedule_cycle(const GrowableArray<VTransformNode*>& stack,
+                                           const VectorSet& pre_visited,
+                                           const VectorSet& post_visited) const {
+  tty->print_cr("\nVTransformGraph::schedule found a cycle on path (P), vectorization attempt fails.");
+  for (int j = 0; j < stack.length(); j++) {
+    VTransformNode* n = stack.at(j);
+    bool on_path = pre_visited.test(n->_idx) && !post_visited.test(n->_idx);
+    tty->print("  %s ", on_path ? "P" : "_");
+    n->print();
+  }
+}
+#endif
+
+void VTransformGraph::apply() {
+#ifndef PRODUCT
+  if (_is_trace_info || TraceLoopOpts) {
+    tty->print_cr("\nVLoopTransformGraph::apply:");
+    lpt()->dump_head();
+    lpt()->head()->dump();
+  }
+#endif
+
+  assert(cl()->is_main_loop(), "auto vectorization only for main loops");
+  phase()->C->print_method(PHASE_AUTO_VECTORIZE1_BEFORE_APPLY, 4, cl());
+
+  apply_memops_reordering_with_schedule();
+  phase()->C->print_method(PHASE_AUTO_VECTORIZE2_AFTER_REORDER, 4, cl());
+
+  adjust_pre_loop_limit_to_align_main_loop_vectors();
+  apply_vectorization();
+  phase()->C->print_method(PHASE_AUTO_VECTORIZE3_AFTER_APPLY, 4, cl());
+}
+
+#ifndef PRODUCT
+void VTransformApplyStatus::trace(VTransformNode* vtn) const {
+  tty->print("  apply: ");
+  vtn->print();
+  tty->print("    ->   ");
+  if (_node == nullptr) {
+    tty->print_cr("nullptr");
+  } else {
+    _node->dump();
+  }
+}
+#endif
+
+// We call "apply" on every VTransformNode, which replaces the packed scalar nodes
+// with vector nodes.
+void VTransformGraph::apply_vectorization() const {
```

Comment:
Note: see the loop over all packs in `SuperWord::output`, and also the `cl` marking and `SuperWordLoopUnrollAnalysis` is taken from there. That code may deserve some attention/refactoring in the future.